### PR TITLE
Add non-link version of 'Not in Library' Button

### LINKED ISSUE #8492: BUG: reloading the page when clicked on not in library button
### Question
is it the right behaviour to reload the page when clicked on not in library?

### Issue resolution criteria
when it shows a pop-up message instead of reloading the page

### PR DESCRIPTION
**Pull request by ClementineAccount**
Closes #8492 but does so by not rendering the link as clickable, rather than by triggering a pop-up.

This pull request is a partial hotfix for the the listed issue by creating an alternate class version of the 'Not in Library' button that does not allow the page to be reloaded when clicked. This Pull Request, however, was only able to fix it partially. I was only able to get the button to show a 'non hyperlink' version on the left screen but not in the table. Need to improve this PR further and get some help for that before then.

### Technical
This uses the 'secondary_action' check in order to verify whether the link should be rendered. It makes the assumption that if 'secondary_action' is enabled, we do not have to render the link.

```
 $# We assume that we are on the Book Page if secondary_action is true  

  $ key = doc.key if is_edition else work_key
  $ is_rendering_hyperlink = not secondary_action
  $:macros.NotInLibrary(key, is_rendering_hyperlink)
```

### Testing

(View video for demonstration)
1. Use the book on Docker at the following link http://localhost:8080/books/OL9M/Ra%CC%84gare%CC%84khalu
2. Verify that buttons on the book page have the 'cannot click' pointer and do not reload the page
3. Search for the book using the search bar.

### Video Demonstration and Screenshot

https://github.com/internetarchive/openlibrary/assets/26779639/b3bd5286-3e12-4021-95f0-0833c3fedfa9

image
(Expected behavior for 'no reload')

One issue I am having is being able to set the 'cannot link' trigger for this button. I tried to check for 'is_edition' but that caused the search page to also have a non-link version, which is a worse bug.

image
I think this button also should not be rendered as a link. But I am unable to figure out how to do that.

### Stakeholders

@scottbarnes helped mentor me for this problem and provided a near complete solution on Slack when I seek help. I still took the initiative of the earlier attempt and testing out the fix. I also chose the idea of using the 'not-allowed' pointer.

### CODE REVIEW

**Comment by ClementineAccount** [comment]
**Additional comment for clarification**

I wish to fix this issue where the link is rendered for the table version of the edition first before merging if possible but I don't know how to do so. I think I need to retrieve the 'page I am currently on' but I am still unfamiliar with what this table's version of the button is called. I attempted 'is_edition' and 'secondary_action' to figure it out but neither approach helped. 

282718835-132974c4-dd2f-428e-bcc5-6279681e9844

**Comment by scottbarnes** [comment]
Thanks for this, @ClementineAccount. I think it's looking really good.

@ClementineAccount , is this a fair summary of the current questions for someone who hasn't been closely following along? There is a question about whether to have a hyperlink for the "Not in Library" button:
1. in the `edition_datatable` (if I have the name correct);
2. when viewing _that edition_; and
3. when that edition is not in library.

Additionally, does the current strategy of changing the background color of the edition one is currently viewing give 'enough' information to a patron to know that an edition in the `edition_datatable` is the one they are viewing?

This last question may be outside of scope of #8492, so it may need a separate issue, but I think we should take both of these questions to our weekly design call at 9 AM Pacific time each Friday.

Here are some examples to help illustrate the issue.

How things look currently on production, on a book page, where there is _only one_ edition. ([See example](https://openlibrary.org/books/OL25292M/Ra%CC%84gare%CC%84khalu)): 
image

This PR currently removes the hyperlink from "Not in Library" in the upper left under the (missing) cover image, but does _not_ remove the hyperlink from "Not in Library" in the `edition_datatable` (note: this link is a link to the very edition one is viewing).
1. Should that link for the 'Not in Library' button in the `edition_datatable`, in the lower right, be removed?
2. Should it be more obvious that the edition in the `edition_datatable` is the one a patron is currently viewing (put another way, is it _confusing_ that one of the links in the edition table "doesn't work" when it looks like all the others)?


Here's how things look on production, on a book page, where there are _multiple_ editions. ([See example](https://openlibrary.org/works/OL8870179W/Sierra_Club_Journal?edition=key%3A/books/OL8636492M&mode=all#editions-list):
image

Here, the changed background color of the first edition in the `edition_datatable`.

Now consider an _example_ of how this might be changed, where the button that says:
```
   Not in Library
(Currently viewing)
```
is _not_ a hyperlink, and has added text to indicate to the patron that this edition is the one at which they're looking on the book page. (Please ignore the added 'outline' around each edition in the `edition_datatable`--that's a separate thing related to merging editions).
image

Finally, although I think you should wait until we get feedback from the community, to answer your question, @ClementineAccount, about how to remove the hyperlink from the relevant edition in the `edition_datatable`, one way would be to take the value of `render_first` (which should be `True` if and only if it's the edition a patron is viewing) from `openlibrary/templates/books/edition-sort.html`, then pass that through to `LoanStatus.html` around [line 91](https://github.com/internetarchive/openlibrary/blob/c6b354408a5174bbaab814d94d3b2c513ed61b9e/openlibrary/templates/books/edition-sort.html#L91) or so, and from there to pass it to `NotInLibrary.html` to use for conditionally rendering a "Not in Library" button that is styled/has text for display in the `edition_datatable`. If that was too little of an overview, just let me know on Slack and I can help clarify, but again, let's wait for feedback from the community. :)

**Comment by ClementineAccount** [comment]
> Finally, although I think you should wait until we get feedback from the community, to answer your question, @ClementineAccount, about how to remove the hyperlink from the relevant edition in the edition_datatable, one way would be to take the value of render_first (which should be True if and only if it's the edition a patron is viewing) from openlibrary/templates/books/edition-sort.html, then pass that through to LoanStatus.html around [line 91](https://github.com/internetarchive/openlibrary/blob/c6b354408a5174bbaab814d94d3b2c513ed61b9e/openlibrary/templates/books/edition-sort.html#L91) or so, and from there to pass it to NotInLibrary.html to use for conditionally rendering a "Not in Library" button that is styled/has text for display in the edition_datatable. If that was too little of an overview, just let me know on Slack and I can help clarify, but again, let's wait for feedback from the community. :)

Thank you for that. I think it is enough for me to get started and figure out some possible solutions. I might update this PR with that feature before the end of the week for completeness (as both buttons should be non-links)  and my own learning for getting used to OpenLibrary. I do, however, understand that how the final feature and merge should look would depend on the community discussion. Having the 'cannot click' pointer was the easiest placeholder to try (as I noticed that having a pointer that does not have a 'click hand' on a button looked even stranger).

**Comment by scottbarnes** [comment]
@ClementineAccount, for now, we're looking at the following:
- Keep `Preview Only`, `Borrow`, and `Read` as blue buttons, and then, *other than on the book page*, replace the dark gray `Not in Library` with a blue button that says `Check Options`, which simply goes to the book page for *that specific edition*.
- Then, on the book page, make `Not in Library` not a link, for everything.
- A follow on issue + PR might be, for the book page, to make `Not in Library` instead say Check local libraries, and it pulls up a modal that uses Library.Link and WorldCat. But that is something for us to think more about after this is done.

**Comment by ClementineAccount** [comment]
> @ClementineAccount, for now, we're looking at the following:
> 
>     * Keep `Preview Only`, `Borrow`, and `Read` as blue buttons, and then, _other than on the book page_, replace the dark gray `Not in Library` with a blue button that says `Check Options`, which simply goes to the book page for _that specific edition_.

Made progress on that one. Just need to do the second part now.
image

**Comment by scottbarnes** [comment]
@ClementineAccount, the code on `8492/Check-Options-Button` looks good to me, aside from two things, one of which was my fault. Sorry about that. :)

I misread my notes and on the book page, in edition table, it should _only_ be the currently selected edition that shows up with `Not in Library` as an unclickable 'button'. Other editions, if there are any, should also have the blue `Check Options` button.

Also, for now, let's leave the `<a>` tag on line 4 of `NotinLibrary.html` on `8492/Check-Options-Button`. There seems to be some CSS styling and making it a `<p>` tag shrinks the text; that is probably not worth investigating just now. If when we circle back to create a 'better' solution, the issue may simply go away.

Finally, I don't think there is need to create a new PR. You can just push your commits to this one. It's probably always easiest to continue pushing commits here, if only so people can give feedback on specific lines in the code instead of having to describe where to make a change.

**Comment by ClementineAccount** [comment]
> @ClementineAccount, the code on `8492/Check-Options-Button` looks good to me, aside from two things, one of which was my fault. Sorry about that. :)
> 
> I misread my notes and on the book page, in edition table, it should _only_ be the currently selected edition that shows up with `Not in Library` as an unclickable 'button'. Other editions, if there are any, should also have the blue `Check Options` button.
> 
> Also, for now, let's leave the `<a>` tag on line 4 of `NotinLibrary.html` on `8492/Check-Options-Button`. There seems to be some CSS styling and making it a `<p>` tag shrinks the text; that is probably not worth investigating just now. If when we circle back to create a 'better' solution, the issue may simply go away.
> 
> Finally, I don't think there is need to create a new PR. You can just push your commits to this one. It's probably always easiest to continue pushing commits here, if only so people can give feedback on specific lines in the code instead of having to describe where to make a change.

Apologies for the late reply. I was busy with other stuff throughout the weekend. 
I have added the commit for '<a>' tag. I want to test if my idea for 'Only selected edition' works but I need another book that has multiple editions to test it. Will check our slack messages for the suggestion of adding a book from prod.

**Comment by ClementineAccount** [comment]
I updated the Pull Request but I don't know why the merge commit has the changes for [Buttoncta.less](https://github.com/ClementineAccount/openlibrary/commit/63dd7be5216df3e3d173ebb50adc2e408c970015#diff-e0bb6d537bebadfe11eb579cb38b09665095564139f9c95750633f32576d6da4)

This is a video demonstrating the current 'Check Options' solution. I created a few mock editions of the book Flatland and imported the book 'Les'. It has some strange names as I was still learning how to use `curl` to add new editions. @scottbarnes helped with that suggestion for importing the books for testing.      

https://github.com/internetarchive/openlibrary/assets/26779639/ce2d695e-5e78-47de-9a31-668e761675f7

---

I also am confused by my own solution due to the variable names (leftover from the previous commits). I think I was just testing out some ideas on the conditional check to get the 'Check Options' to show properly in both the table and the 'search page' and accidentally fixed it earlier than expected or something like that. 

I think the fact that 'is_rendering_hyperlink' checks if it is not a secondary edition and then checks if its on a book page is because I had to 'override' the logic if it there is an edition but the 'else' was triggered. It does work so far, however, so I have pushed it to the branch for review.

```
$else:
  $# We assume that we are on the Book Page if secondary_action is true  
  $# TODO: Clean up these conditional checks once I verify it works
  $# The second 'else' after if is_edition checking to see if I can use 'is_edition' for table button logic
  $ key = doc.key if is_edition else work_key
  $ is_rendering_hyperlink = not secondary_action
  $if is_rendering_hyperlink and is_book_page:
      $ is_rendering_hyperlink = False
  $if is_rendering_hyperlink:
      $:macros.CheckOptions(key)
  $else:
    $if is_edition:
      $:macros.NotInLibrary(key)
    $else:
      $:macros.CheckOptions(key) 
```

File: openlibrary/macros/NotInLibrary.html
Comment (scottbarnes) [inline openlibrary/macros/NotInLibrary.html:1]:
```suggestion
$def with()
```
Let's remove this parameter as it is now unused.

**Review by scottbarnes — CHANGES_REQUESTED** [review]
I re-wrote the history quite a bit, so I am asking you to review my changes to ensure I didn't remove any changes you intended to have. I'd say this is especially true about `buttonCta.less` which was changed by other commits over the course of this PR, and which was generating a conflict.

The reason for re-writing the history was that some changes from the past version of `buttonCta.less` were overwriting changes made in other PRs. That should be fixed now, but I want to ensure I didn't remove something that should have been there.

If everything looks good you can pull this down locally and work with it more, or accept my one suggestion here and I think we can call this one done. :)

**Review by scottbarnes — APPROVED** [review]
Looks good to me. Thanks for this, @ClementineAccount!

**Pull request merged**
